src/main/cpp/apple_fsnotifier.cpp

```diff
@@ -222,56 +226,65 @@ JNIEnv* Server::getThreadEnv() {
     return lookupThreadEnv(jvm);
 }
 
-JNIEXPORT jobject JNICALL
-Java_net_rubygrapefruit_platform_internal_jni_OsxFileEventFunctions_startWatching(JNIEnv *env, jclass target, jobjectArray paths, long latencyInMillis, jobject javaCallback, jobject result) {
-
-    log_fine(env, "Configuring...", NULL);
-
+Server *startWatching(JNIEnv *env, jclass target, jobjectArray paths, long latencyInMillis, jobject javaCallback) {
     int count = env->GetArrayLength(paths);
     if (count == 0) {
-        log_severe(env, "No paths given to watch", NULL);
-        // TODO Error handling
-        return NULL;
+        throw FileWatcherException("No paths given to watch");
     }
 
     CFMutableArrayRef rootsToWatch = CFArrayCreateMutable(NULL, count, NULL);
     if (rootsToWatch == NULL) {
-        log_severe(env, "Could not allocate array to store roots to watch", NULL);
-        // TODO Error handling
-        return NULL;
+        throw FileWatcherException("Could not allocate array to store roots to watch");
     }
 
     for (int i = 0; i < count; i++) {
         jstring path = (jstring) env->GetObjectArrayElement(paths, i);
-        char* watchedPath = java_to_char(env, path, result);
-        log_fine(env, "Watching %s", watchedPath);
+        char* watchedPath = java_to_char(env, path, NULL);
         if (watchedPath == NULL) {
-            log_severe(env, "Could not allocate string to store root to watch.", NULL);
-            // TODO Free resources
-            return NULL;
+            throw FileWatcherException("Could not allocate string to store root to watch");
```

Comment:
Don't we need to free resources when the exception is thrown?

Reply:
That part is not done yet. I am actively learning about `unique_ptr` and friends to make sure they get released.

Reply:
I think we are actually much better now wrt freeing resources when a problem occurs.